### PR TITLE
feat: better debug support

### LINKED ISSUE #185: Support Custom Launch Configurations
**Is your feature request related to a problem? Please describe.**
I am authoring a cloudflare worker with an end-to-end test that starts `workerd` on a custom inspect port. When clicking the "debug" button, vitest uses its own custom debug launch config and I have no way to set up to attach to the `workerd` instance in order to debug internals.

**Describe the solution you'd like**
Provide a configuration option to supply a custom debug launch config, similar to the `Mocha Test Explorer`. vitest would then append its -t, --api-port, and --api-host to the args of that launch config.

Alternatively, provide an additional attach port setting in the config that will launch a child attach debug config on the specified port.

I would be willing to attempt a PR if there is interest, I wrote my own Test Extension for PowerShell Pester and am familiar with the internals.

### LINKED ISSUE #320: Extension hangs if debugger is detached during a breakpoint pause
### Describe the bug

After running a test in debug mode 

### Reproduction

Can't really reproduce on StackBlitz. I can reproduce it on the [basic example](https://github.com/vitest-dev/vitest/tree/main/examples/basic). The only modification is my `vite.config.ts` specifies this:
```
  test: {
    include: [
      "app/**/*.test.server.ts",
    ],
```

1. Debug a test with a breakpoint
image
2. Once paused on the breakpoint, hit either "Disconnect" or "Stop" from the debug panel
image
3. Notice the test keeps the spinning running icon:
image
4. Right click the same test -> Debug test, this works and it can be re-run
5. Right click -> Run test (*NOT debug*) (or run test from test explorer view)
image
5. Notice it never runs
image

6. Right click the same test -> Debug test, this no longer works either and they just keep queuing up 

This also seems to happen if the test being debugged fails.

I am unable to run _any_ other test using the regular run config after that without reloading the editor window.

<details><summary>vitest extension logs</summary>
<p>

[INFO 12:54:29 AM] [Vitest] Extension is activated because Vitest is installed or there is a Vite/Vitest config file in the workspace.
[INFO 12:54:29 AM] [API] Running Vitest: v1.4.0 (/Volumes/projects/modular-bundles/extensions/modular-bundles/vite.config.js), v1.4.0 (/Volumes/projects/modular-bundles/vite.config.ts)
[INFO 12:54:31 AM] [API] Vitest process 79597 created
[INFO 12:54:34 AM] Running 1 file(s) with name pattern: ^\s?findMigrations finds all files$
[Worker] Debugger listening on ws://127.0.0.1:9229/347ba504-6379-4364-99d0-40b58524e7e8
For help, see: https://nodejs.org/en/docs/inspector
[Worker] Debugger attached.
[INFO 12:54:38 AM] [DEBUG] Debugging started
[INFO 12:54:38 AM] Running 1 file(s) with name pattern: ^\s?findMigrations finds all files$
[INFO 12:54:39 AM] Running 1 file(s) with name pattern: ^\s?findMigrations finds all files$
[Worker] Debugger ending on ws://127.0.0.1:9229/347ba504-6379-4364-99d0-40b58524e7e8
For help, see: https://nodejs.org/en/docs/inspector
[INFO 12:57:04 AM] Running 1 file(s) with name pattern: ^\s?findMigrations finds all files$
[INFO 12:57:04 AM] [API] Running Vitest: v1.4.0 (/Volumes/projects/modular-bundles/extensions/modular-bundles/vite.config.js), v1.4.0 (/Volumes/projects/modular-bundles/vite.config.ts)
[INFO 12:57:05 AM] [API] Vitest process 80839 created
[Worker] Debugger listening on ws://127.0.0.1:9229/2faac4d4-5ac3-4cdd-ba0d-3766f4a55a6f
For help, see: https://nodejs.org/en/docs/inspector
[Worker] Debugger attached.
[INFO 12:57:19 AM] [DEBUG] Debugging started
[INFO 12:57:19 AM] Running 1 file(s) with name pattern: ^\s?findMigrations finds all files$
[INFO 12:57:20 AM] Running 1 file(s) with name pattern: ^\s?findMigrations finds all files$
[Worker] Debugger ending on ws://127.0.0.1:9229/2faac4d4-5ac3-4cdd-ba0d-3766f4a55a6f
For help, see: https://nodejs.org/en/docs/inspector
[INFO 12:57:26 AM] Running 1 file(s) with name pattern: ^\s?findMigrations finds all files$
[Worker] Debugger listening on ws://127.0.0.1:9229/ad9d6116-c4a3-4f95-af6b-38c7f43a1f85
For help, see: https://nodejs.org/en/docs/inspector
[Worker] Debugger attached.
[INFO 12:57:43 AM] [DEBUG] Debugging started
[INFO 12:57:44 AM] Running 1 file(s) with name pattern: ^\s?findMigrations finds all files$
[INFO 12:57:44 AM] Running 1 file(s) with name pattern: ^\s?findMigrations finds all files$
[Worker] Debugger ending on ws://127.0.0.1:9229/ad9d6116-c4a3-4f95-af6b-38c7f43a1f85
For help, see: https://nodejs.org/en/docs/inspector
[Error 1:00:10 AM] [API] [Error Error] Channel closed
Error: Channel closed
    at new NodeError (node:internal/errors:405:5)
    at ChildProcess.send (node:internal/child_process:740:16)
    at post (/Users/xavier/.vscode/extensions/vitest.explorer-0.5.9/dist/extension.js:3:2594)
    at /Users/xavier/.vscode/extensions/vitest.explorer-0.5.9/dist/extension.js:3:1068
    at new Promise (<anonymous>)
    at Proxy.b (/Users/xavier/.vscode/extensions/vitest.explorer-0.5.9/dist/extension.js:3:823)

</p>
</details> 

### System Info

```shell
System:
    OS: macOS 14.1.1
    CPU: (12) arm64 Apple M2 Pro
    Memory: 92.25 MB / 16.00 GB
    Shell: 3.7.0 - /opt/homebrew/bin/fish
  Binaries:
    Node: 18.19.0 - ~/.asdf/installs/nodejs/18.19.0/bin/node
    Yarn: 1.22.21 - /Volumes/projects/modular-bundles/node_modules/.bin/yarn
    npm: 9.9.3 - /Volumes/projects/modular-bundles/node_modules/.bin/npm
  IDEs:
    VSCode: 1.87.2 - /usr/local/bin/code
    Vim: 9.0 - /usr/bin/vim
    Xcode: /undefined - /usr/bin/xcodebuild
  Browsers:
    Chrome: 123.0.6312.59
    Safari: 17.1
```
```


### Used Package Manager

yarn

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vscode/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vscode/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vscode/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vscode/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by sheremet-va**
This PR improves the debugger support by utilizing the new [`inspector` option](https://github.com/vitest-dev/vitest/pull/5509). Also, a new option was added: `vitest.debuggerAddress`. With the `vitest.debuggerPort`, I think we can close #185 now.

Fixes #320
The debugger requires Vitest 1.5.0, but all other features work as they did with 1.4.0

### CODE REVIEW

**Comment by ffMathy** [comment]
Would introducing some kind of state machine pattern help? I feel like it's one of those problems that could be solved by this.

**Comment by sheremet-va** [comment]
> Would introducing some kind of state machine pattern help?

Can you be more specific?

**Comment by ffMathy** [comment]
Well, let's say that when pressing a particular button (like "start") has different effects in different states.

Maybe when we are in the ContinuousRun state, that button does something else than when we are in another state.

We can represent each state as a class implementing a particular interface with various inputs (onStartClicked) etc, and also allowing each state to control what other states are allowed to be switched to and from.

It's similar to the strategy pattern in a way. I don't know the specifics of the problem, but it sounds like you're describing complex flows that apply differently in particular states.

Can you describe the flows more detailed? Then I can come up with something probably.

Alternatively, for me, it helps drawing it as a state machine on a piece of paper.

**Comment by ffMathy** [comment]
More info: https://en.m.wikipedia.org/wiki/UML_state_machine

**Comment by ffMathy** [comment]
And the pattern (you can create variants of it though): https://refactoring.guru/design-patterns/state

**Comment by sheremet-va** [comment]
> We can represent each state as a class implementing a particular interface with various inputs

I see what you mean now, but the problem here is not that we can't organize it, but that it's not possible to do several things at once - for example, you can't debug a test and collect other test files at the same time because Vitest only supports a single pool (it blocks other test runs until the current one is completed).

**Comment by ffMathy** [comment]
Then each state class can have a property called "canDebug" etc, to represent that.

**Comment by sheremet-va** [comment]
> Then each state class can have a property called "canDebug" etc, to represent that.

Again, this has nothing to do with the code architecture. We _need_ to be able to collect tests _despite_ the debugging.

**Comment by ffMathy** [comment]
Ah I see. My bad.

File: src/debug/debugManager.ts [hunk 1]
```diff
@@ -1,19 +1,16 @@
...
-  runner: TestRunner
-  api: VitestFolderAPI
-}
 
 export class TestDebugManager extends vscode.Disposable {
   private disposables: vscode.Disposable[] = []
   private sessions = new Set<vscode.DebugSession>()
-  private configurations = new Map<string, VitestDebugConfig>()
+  private port: number | undefined
+  private address: string | undefined
```
Comment (ffMathy) [inline src/debug/debugManager.ts:11]:
```suggestion
  private port?: number
  private address?: string
```

File: src/worker/actions.ts [hunk 2]
```diff
@@ -1,265 +1,32 @@
...
+  return new Proxy<any>({
     async close() {
-      for (const vitest of vitestEntries) {
+      for (const vitest in vitestById) {
         try {
-          await vitest[1].close()
+          await vitestById[vitest].dispose()
         }
         catch {
           // ignore
```
Comment (ffMathy) [inline src/worker/actions.ts:12]: When can this happen? Is it worth adding logging here? Or guard against it instead, and "seal it off" with tests?

File: src/worker/actions.ts [hunk 3]
```diff
@@ -1,265 +1,32 @@
...
         }
       }
     },
-  }
+  }, {
+    get(target, prop) {
```
Comment (ffMathy) [inline src/worker/actions.ts:17]: Why is a proxy needed here? Perhaps a comment explaining it would make sense?

**Review by ffMathy — APPROVED** [review]
Beautiful. Great work! Minor optional suggestions and preferences.

File: src/debug/debugManager.ts (hunk 1 above)
Comment (sheremet-va) [reply src/debug/debugManager.ts:11]: I prefer explicit `undefined`, it's not optional, we defined it as undefined

File: src/worker/actions.ts (hunk 2 above)
Comment (sheremet-va) [reply src/worker/actions.ts:12]: https://github.com/vitest-dev/vitest/issues/2008

File: src/debug/debugManager.ts (hunk 1 above)
Comment (ffMathy) [reply src/debug/debugManager.ts:11]: Fair enough 🙂

**Pull request merged**